src/main.ts

```diff
@@ -199,15 +203,19 @@ async function injectViteIndexMiddleware(
   app.get("/*", async (req, res, next) => {
     if (isIgnoredPath(req.path, req)) return next();
 
-    const template = fs.readFileSync(
-      path.resolve(config.root, "index.html"),
-      "utf8"
-    );
-
     if (isStaticFilePath(req.path)) next();
     else {
-      const html = await server.transformIndexHtml(req.originalUrl, template);
-      res.send(getTransformedHTML(html, req));
+      try {
+        const basePath = getBasePath(req.path);
+        const template = fs.readFileSync(
+          path.join(config.root, basePath, "index.html"),
+          "utf8"
+        );
```

Comment:
Right now if the file doesn't exist, an error is thrown and that breaks client-side routing.
When you go to `/some/path/nested`, and file that it looks for (for example `/some/path/index.html`) doesn't exist, it should fallback to `/some/index.html` and if that also doesn't exist, the root `index.html`.

That way you can have:
- `/admin/my/secret/route` -> `/admin/index.html` is displayed and client-side routing is used to render correct components for `/my/secret/route`
- `/whatever/other/route` -> `index.html` is served

I suggest creating a function that will try to sequentially find a nested index html file for a route.
For the `/admin/my/secret/route` path the process would look like this:
- `admin/my/secret/index.html`
- `admin/my/index.html`
- `admin/index.html`
- `index.html`

So these files would be selected from top to bottom, if one doesn't exist, try to select another one. This behaviour seems to mimic what Vite does and also keeps client-side routing intact.